stacks-common/src/util/secp256k1.rs

```diff
@@ -0,0 +1,1016 @@
+}
+
+impl SignatureCompat {
+    /// Serializes the signature in DER format.
+    pub fn serialize_der(&self) -> Vec<u8> {
+        self.signature.to_der().as_bytes().to_vec()
+    }
+}
+
+impl From<(K256Signature, K256RecoveryId)> for RecoverableSignature {
+    fn from((signature, recovery_id): (K256Signature, K256RecoveryId)) -> Self {
+        RecoverableSignature {
+            signature,
+            recovery_id,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct Secp256k1PublicKey {
+    // serde is broken for secp256k1, so do it ourselves
+    #[serde(
+        serialize_with = "secp256k1_pubkey_serialize",
+        deserialize_with = "secp256k1_pubkey_deserialize"
+    )]
+    key: K256VerifyingKey,
+    compressed: bool,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct Secp256k1PrivateKey {
+    // serde is broken for secp256k1, so do it ourselves
+    #[serde(
+        serialize_with = "secp256k1_privkey_serialize",
+        deserialize_with = "secp256k1_privkey_deserialize"
+    )]
+    key: K256SigningKey,
+    compress_public: bool,
+}
+
+impl Hash for Secp256k1PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Hash based on the compressed public key bytes for consistency
+        self.to_bytes_compressed().hash(state);
+    }
+}
+
+impl MessageSignature {
+    /// Creates an "empty" signature (all zeros). Note this is not a valid signature.
+    pub fn empty() -> MessageSignature {
+        // NOTE: this cannot be a valid signature
+        MessageSignature([0u8; 65])
+    }
+
+    /// Generates place-holder data (for testing purposes only).
+    #[cfg(any(test, feature = "testing"))]
+    pub fn from_raw(sig: &[u8]) -> MessageSignature {
+        let mut buf = [0u8; 65];
+        if sig.len() < 65 {
+            buf[..sig.len()].copy_from_slice(sig);
+        } else {
+            buf.copy_from_slice(&sig[..65]);
+        }
+        MessageSignature(buf)
+    }
+
+    /// Converts from a secp256k1::ecdsa::RecoverableSignature to our MessageSignature.
+    pub fn from_secp256k1_recoverable(sig: &RecoverableSignature) -> MessageSignature {
+        let (recid, bytes) = sig.serialize_compact();
+        let mut ret_bytes = [0u8; 65];
+        let recovery_id_byte = recid.to_byte();
+        ret_bytes[0] = recovery_id_byte;
+        ret_bytes[1..=64].copy_from_slice(&bytes[..64]);
+        MessageSignature(ret_bytes)
+    }
+
+    /// Converts to a secp256k1::ecdsa::RecoverableSignature.
+    pub fn to_secp256k1_recoverable(&self) -> Option<RecoverableSignature> {
+        let recovery_id = K256RecoveryId::from_byte(self.0[0])?;
+        let signature = K256Signature::from_slice(&self.0[1..65]).ok()?;
+        Some(RecoverableSignature {
+            signature,
+            recovery_id,
+        })
+    }
+
+    /// Converts from VRS to RSV.
+    pub fn to_rsv(&self) -> Vec<u8> {
+        [&self.0[1..], &self.0[0..1]].concat()
```

Comment:
😬 not sure of a cleaner one here
```suggestion
        let (v, rest) = self
            .0
            .split_first()
            .expect("MessageSignature must contain at least one byte");
        let mut out = rest.to_vec();
        out.push(*v);
        out
```